updater/updater.go

```
@@ -0,0 +1,339 @@
+// and sets the version in the UpdateInfo as the current version of
+// the instance in the updater.
+func (u *updater) CompleteUpdate(ctx context.Context, info UpdateInfo) error {
+	if info == nil {
+		return errors.New("invalid UpdateInfo")
+	}
+
+	version := info.Version()
+	if version == "" {
+		return fmt.Errorf("invalid version, can't report complete event to omaha server")
+	}
+
+	err := u.ReportProgress(ctx, ProgressUpdateComplete)
+	if err != nil {
+		return fmt.Errorf("reporting ProgressUpdateComplete to omaha server: %w", err)
+	}
+
+	u.SetInstanceVersion(version)
+	return nil
+}
+
+// SendOmahaRequest sends the event request to the omaha server
+// and returns the omaha.Response returned by the omaha server.
+func (u *updater) SendOmahaEvent(ctx context.Context, event *omaha.EventRequest) (*omaha.Response, error) {
+	req := u.newAppRequest()
+	app := req.GetApp(u.appID)
+	app.Events = append(app.Events, event)
+
+	return u.SendOmahaRequest(ctx, req)
+}
+
+// InstanceVersion returns the current version of the application.
+func (u *updater) InstanceVersion() string {
+	u.mu.RLock()
+	defer u.mu.RUnlock()
+	return u.instanceVersion
+}
+
+// SetInstanceVersion sets the current instance version
+// of the application to the updater.
+func (u *updater) SetInstanceVersion(version string) {
+	u.mu.Lock()
+	defer u.mu.Unlock()
+	u.instanceVersion = version
+}
+
+// TryUpdate function takes in an UpdateHandler and performs
+// the complete flow from checking for updates to reporting
+// status etc and returns an error if anything fails in the flow.
+func (u *updater) TryUpdate(ctx context.Context, handler UpdateHandler) error {
+	if handler == nil {
+		return errors.New("invalid UpdateHandler")
+	}
+
+	info, err := u.CheckForUpdates(ctx)
+	if err != nil {
+		return fmt.Errorf("checking for update: %w", err)
+	}
+
+	if !info.HasUpdate() {
+		return fmt.Errorf("no update available for app %v, channel %v: %v", u.appID, u.channel, info.UpdateStatus())
```

Should no update be treated as error? If so, perhaps it should be a typed error, so user can use `errors.Is()` to distinguish between no updates and failed update?

---

This was something I thought about initially. But for me, the concept of getting an error when there's no update available can be misleading. And having to check use `errors.Is` for understanding whether there's an error is a bit less readable than having `info.HasUpdate()` IMO.

Looking at it again, if we are not doing anything  with the struct when there's no update available (besides checking if there's an update), maybe it's more "goist" to do what you're saying.

@yolossn , WDYT?

---

Since the `TryUpdate` function tries to automatically checks and applies the `UpdateHandler` functions, If an error is not returned for this the user will not know the difference between successfully update being applied and skipping the steps as no update is available. Note that the `UpdateInfo` is not available to the user to check `info.HasUpdate()`.

Also, the goal of `TryUpdate` is to abstract as much Omaha-specific knowledge as possible, I prefer returning a typed error.